File: scheduler.js
#!/usr/bin/env node

import SpacePipeline from './pipeline.js';
import fs from 'fs';

class VideoScheduler {
  constructor() {
    this.lastRunFile = 'last_run.json';
  }

  saveLastRun(results) {
    const runInfo = {
      timestamp: new Date().toISOString(),
      status: results.status,
      duration: results.total_duration_seconds,
      video_url: results.final_outputs?.video_url
    };
    
    fs.writeFileSync(this.lastRunFile, JSON.stringify(runInfo, null, 2));
  }

  getLastRunInfo() {
    try {
      const data = fs.readFileSync(this.lastRunFile, 'utf8');
      return JSON.parse(data);
    } catch (error) {
      return null;
    }
  }

  shouldRunNow() {
    const lastRun = this.getLastRunInfo();
    if (!lastRun) return true;
    
    const lastRunTime = new Date(lastRun.timestamp);
    const timeSinceLast = Date.now() - lastRunTime.getTime();
    const twoHours = 2 * 60 * 60 * 1000; // 2 hours in milliseconds
    
    if (timeSinceLast < twoHours && lastRun.status === 'completed') {
      console.log(`Skipping run - last successful run was ${Math.round(timeSinceLast / (60 * 1000))} minutes ago`);
      return false;
    }
    
    return true;
  }

  async runScheduledVideoGeneration() {
    console.log('🕐 Scheduled video generation starting...');
    
    if (!this.shouldRunNow()) {
      return;
    }
    
    try {
      const pipeline = new SpacePipeline();
      const results = await pipeline.runPipeline();
      
      this.saveLastRun(results);
      
      if (results.status === 'completed') {
        const videoUrl = results.final_outputs?.video_url;
        console.log('✅ Scheduled video generation completed successfully!');
        if (videoUrl) {
          console.log(`🎬 Video URL: ${videoUrl}`);
        }
      } else {
        console.error('❌ Scheduled video generation failed');
      }
      
    } catch (error) {
      console.error(`Scheduled run failed: ${error.message}`);
    }
  }

  async startScheduler() {
    console.log('🚀 Video generation scheduler started');
    console.log('Running every 4 hours...');
    
    // Run immediately if no recent run
    if (!this.getLastRunInfo()) {
      console.log('No previous runs found, running immediately...');
      await this.runScheduledVideoGeneration();
    }
    
    // Schedule to run every 4 hours (14400000 ms)
    setInterval(async () => {
      await this.runScheduledVideoGeneration();
    }, 4 * 60 * 60 * 1000);
    
    console.log('Scheduler is running. Press Ctrl+C to stop.');
  }
}

async function main() {
  const scheduler = new VideoScheduler();
  await scheduler.startScheduler();
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch(console.error);
}

export default VideoScheduler;

File: scheduler.py
#!/usr/bin/env python3
"""
Scheduler for automated video generation
"""
import os
import sys
import time
import logging
import schedule
import json
from datetime import datetime, timedelta

# Add pipeline to path
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from pipeline import WorkflowOrchestrator

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('scheduler.log'),
        logging.StreamHandler(sys.stdout)
    ]
)

logger = logging.getLogger(__name__)

class VideoScheduler:
    def __init__(self):
        self.orchestrator = WorkflowOrchestrator()
        self.last_run_file = 'last_run.json'
        
    def save_last_run(self, results):
        """Save information about the last run"""
        run_info = {
            'timestamp': datetime.now().isoformat(),
            'status': results.get('status'),
            'duration': results.get('total_duration_seconds'),
            'video_url': results.get('final_video', {}).get('video_url')
        }
        
        with open(self.last_run_file, 'w') as f:
            json.dump(run_info, f, indent=2)
    
    def get_last_run_info(self):
        """Get information about the last run"""
        try:
            with open(self.last_run_file, 'r') as f:
                return json.load(f)
        except FileNotFoundError:
            return None
    
    def should_run_now(self):
        """Check if we should run based on last run time"""
        last_run = self.get_last_run_info()
        if not last_run:
            return True
        
        last_run_time = datetime.fromisoformat(last_run['timestamp'])
        time_since_last = datetime.now() - last_run_time
        
        # Don't run if last run was less than 2 hours ago and successful
        if time_since_last < timedelta(hours=2) and last_run.get('status') == 'completed':
            logger.info(f"Skipping run - last successful run was {time_since_last} ago")
            return False
        
        return True
    
    def run_scheduled_video_generation(self):
        """Run the video generation workflow"""
        logger.info("Scheduled video generation starting...")
        
        if not self.should_run_now():
            return
        
        try:
            # Health check first
            health_status = self.orchestrator.health_check_all_services()
            failed_services = [service for service, status in health_status.items() if not status]
            
            if failed_services:
                logger.warning(f"Some services are not healthy: {failed_services}")
                # Continue anyway - some services might recover
            
            # Run workflow with different configurations based on time of day
            current_hour = datetime.now().hour
            
            if 6 <= current_hour < 12:  # Morning
                config = {'duration': 90, 'voice': 'nova'}
            elif 12 <= current_hour < 18:  # Afternoon  
                config = {'duration': 120, 'voice': 'alloy'}
            else:  # Evening/Night
                config = {'duration': 150, 'voice': 'echo'}
            
            logger.info(f"Running workflow with config: {config}")
            results = self.orchestrator.run_scheduled_workflow(config)
            
            # Save results
            self.save_last_run(results)
            
            if results['status'] == 'completed':
                video_url = results.get('final_video', {}).get('video_url')
                logger.info(f"✅ Video generation completed successfully!")
                if video_url:
                    logger.info(f"🎬 Video URL: {video_url}")
                    
                    # Here you could add integration with YouTube API to upload
                    # or send notifications, etc.
                    
            else:
                logger.error(f"❌ Video generation failed: {results.get('errors', [])}")
                
        except Exception as e:
            logger.error(f"Scheduled run failed with exception: {e}")

def main():
    scheduler = VideoScheduler()
    
    # Schedule video generation
    # Run every 4 hours during active hours
    schedule.every(4).hours.do(scheduler.run_scheduled_video_generation)
    
    # Also run at specific times for better scheduling
    schedule.every().day.at("08:00").do(scheduler.run_scheduled_video_generation)
    schedule.every().day.at("14:00").do(scheduler.run_scheduled_video_generation)
    schedule.every().day.at("20:00").do(scheduler.run_scheduled_video_generation)
    
    logger.info("Video generation scheduler started")
    logger.info("Scheduled times: Every 4 hours, plus 8:00, 14:00, and 20:00 daily")
    
    # Run once immediately if no recent run
    if not scheduler.get_last_run_info():
        logger.info("No previous runs found, running immediately...")
        scheduler.run_scheduled_video_generation()
    
    # Keep the scheduler running
    try:
        while True:
            schedule.run_pending()
            time.sleep(60)  # Check every minute
    except KeyboardInterrupt:
        logger.info("Scheduler stopped by user")

if __name__ == "__main__":
    main()